core/mcp.go

```diff
@@ -847,101 +783,317 @@ func (m *MCP) Builtins(srv *dagql.Server, tools []LLMTool) ([]LLMTool, error) {
-						continue
-					}
-					desc += "\n- " + tool.Name + " (returns " + tool.Returns + ")"
-				}
-				var objects []string
-				for _, typeName := range slices.Sorted(maps.Keys(m.env.typeCounts)) {
-					count := m.env.typeCounts[typeName]
-					for i := 1; i <= count; i++ {
-						bnd := m.env.objsByID[fmt.Sprintf("%s#%d", typeName, i)]
-						objects = append(objects, fmt.Sprintf("%s: %s", bnd.ID(), bnd.Description))
-					}
+	builtins = append(builtins, LLMTool{
+		Name:        "list_objects",
+		Description: "List available objects.",
+		Schema: map[string]any{
+			"type":                 "object",
+			"properties":           map[string]any{},
+			"required":             []string{},
+			"additionalProperties": false,
+		},
+		Strict: true,
+		Call: ToolFunc(srv, func(ctx context.Context, args struct{}) (any, error) {
+			type objDesc struct {
+				ID          string `json:"id"`
+				Description string `json:"description"`
+			}
+			var objects []objDesc
+			for _, typeName := range slices.Sorted(maps.Keys(m.env.typeCounts)) {
+				count := m.env.typeCounts[typeName]
+				for i := 1; i <= count; i++ {
+					bnd := m.env.objsByID[fmt.Sprintf("%s#%d", typeName, i)]
+					objects = append(objects, objDesc{
+						ID:          bnd.ID(),
+						Description: bnd.Description,
+					})
 				}
-				if len(objects) > 0 {
-					desc += "\n\nAvailable objects:"
-					for _, input := range objects {
-						desc += fmt.Sprintf("\n- %s", input)
+			}
+			return toolStructuredResponse(objects)
+		}),
+	})
+
+	builtins = append(builtins, LLMTool{
+		Name:        "list_methods",
+		Description: "List the methods that can be selected.",
+		Schema: map[string]any{
+			"type":                 "object",
+			"properties":           map[string]any{},
+			"required":             []string{},
+			"additionalProperties": false,
+		},
+		Strict: true,
+		Call: ToolFunc(srv, func(ctx context.Context, args struct{}) (any, error) {
+			type toolDesc struct {
+				Name         string            `json:"name"`
+				Returns      string            `json:"returns"`
+				RequiredArgs map[string]string `json:"required_args,omitempty"`
+			}
+			var methods []toolDesc
+			for _, method := range allMethods {
+				reqArgs := map[string]string{}
+				var returns string
+				if method.Field != nil {
+					returns = method.Field.Type.String()
+					for _, arg := range method.Field.Arguments {
+						if arg.DefaultValue != nil || !arg.Type.NonNull {
+							// optional
+							continue
+						}
+						reqArgs[arg.Name] = arg.Type.String()
 					}
 				}
-				return desc
-			})(),
+				methods = append(methods, toolDesc{
+					Name:         method.Name,
+					RequiredArgs: reqArgs,
+					Returns:      returns,
+				})
+			}
+			sort.Slice(methods, func(i, j int) bool {
+				return methods[i].Name < methods[j].Name
+			})
+			return toolStructuredResponse(methods)
+		}),
+	})
+
+	if len(allMethods) > 0 {
+		builtins = append(builtins, LLMTool{
+			Name:        "select_methods",
```

Comment:
maybe hot take, but if we have list_methods and call_methods, why do we need select_methods?

Reply:
* `list_methods` returns a broad and brief view of all available methods, WITHOUT descriptions and ONLY showing required args
* `select_methods` returns the fully detailed view of the methods you requested, WITH descriptions and the FULL schema

If we merged them, there would be a massive increase in token cost and possibly overwhelm the context window.

Reply:
cool, valid.